MapleServer2/Tools/InventoryController.cs

```diff
@@ -1,46 +1,153 @@
                 if (i.Id != item.Id || i.Amount >= i.SlotMax)
                 {
                     continue;
                 }
-                // Updates inventory for item amount overflow
+                // Updates inventory for item amount overflow.
                 if ((i.Amount + item.Amount) > i.SlotMax)
                 {
-                    int added = i.SlotMax - i.Amount; // For marking item new with correct added amount
-
                     item.Amount = item.Amount - (i.SlotMax - i.Amount);
                     i.Amount = i.SlotMax;
                     session.Send(ItemInventoryPacket.Update(i.Uid, i.Amount));
-                    session.Send(ItemInventoryPacket.MarkItemNew(i, added));
                 }
                 // Updates item amount
                 else
                 {
                     i.Amount = i.Amount + item.Amount;
                     session.Send(ItemInventoryPacket.Update(i.Uid, i.Amount));
-                    session.Send(ItemInventoryPacket.MarkItemNew(i, item.Amount));
                     return;
                 }
             }
         }
-        // Add item to inventory if cannot stack any further
-        session.Player.Inventory.Add(item); 
-        session.Send(ItemInventoryPacket.Add(item));
-        session.Send(ItemInventoryPacket.MarkItemNew(item, item.Amount));
+        session.Player.Inventory.Add(item); // adds item into internal database
+        session.Send(ItemInventoryPacket.Add(item)); // sends packet to add item clientside.
+
+        if (isNew)
+        {
+            session.Send(ItemInventoryPacket.MarkItemNew(item, item.Amount)); // Marks Item as New
+        }
+    }
+
+    // Removes Item from inventory by reference
+    public static void Remove(GameSession session, out Item item, long uid)
+    {
+        session.Player.Inventory.Remove(uid, out item);
+        session.Send(ItemInventoryPacket.Remove(uid));
+    }
+
+    // Removes Item from inventory
+    public static void Remove(GameSession session, Item item, long uid)
+    {
+        session.Player.Inventory.Remove(uid, out item);
+        session.Send(ItemInventoryPacket.Remove(uid));
+    }
+
+    // Picks up item
+    public static void PickUp(GameSession session, Item item)
+    {
+        session.Player.Inventory.Add(item); // adds item into internal database
+        session.Send(ItemInventoryPacket.Add(item)); // sends packet to add item clientside.
+    }
+
+    // Drops item with option to drop bound items.
+    public static void DropItem(GameSession session, long uid, int amount, Boolean isbound)
+    {
+        if (!isbound) // Drops Item.
+        {
+            int remaining = session.Player.Inventory.Remove(uid, out Item droppedItem, amount); // returns remaining amount of item.
+            if (remaining < 0)
+            {
+                return; // Removal failed
+            }
+
+            // Updates item amount and removes item.
+            session.Send(remaining > 0
```

Comment:
Ternary operator shouldn't be used here since it's sending entirely different packets